File: packages/python/src/jsonld_ex/validation.py
"""Validation Extensions for JSON-LD (@shape)."""

from __future__ import annotations
import re
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any, Optional, Sequence


@dataclass
class ValidationError:
    path: str
    constraint: str
    message: str
    value: Any = None


@dataclass
class ValidationWarning:
    path: str
    code: str
    message: str


@dataclass
class ValidationResult:
    valid: bool
    errors: list[ValidationError] = field(default_factory=list)
    warnings: list[ValidationWarning] = field(default_factory=list)


XSD = "http://www.w3.org/2001/XMLSchema#"


def validate_node(node: dict[str, Any], shape: dict[str, Any]) -> ValidationResult:
    """Validate a JSON-LD node against a shape definition."""
    errors: list[ValidationError] = []
    warnings: list[ValidationWarning] = []

    if not isinstance(node, dict):
        errors.append(ValidationError(".", "type", "Node must be a dict"))
        return ValidationResult(False, errors, warnings)

    # Type check
    if "@type" in shape:
        node_types = _get_types(node)
        if shape["@type"] not in node_types:
            errors.append(ValidationError(
                "@type", "type",
                f'Expected type "{shape["@type"]}", found: {node_types}',
                node_types,
            ))

    # Property constraints
    for prop, constraint in shape.items():
        if prop.startswith("@") or not isinstance(constraint, dict):
            continue

        value = node.get(prop)

        # ── Cardinality constraints (operate on raw value before extraction) ──
        count = _count_values(value)

        if "@minCount" in constraint:
            min_count = constraint["@minCount"]
            if count < min_count:
                errors.append(ValidationError(
                    prop, "minCount",
                    f"Expected at least {min_count} value(s), found {count}",
                    value,
                ))

        if "@maxCount" in constraint:
            max_count = constraint["@maxCount"]
            if count > max_count:
                errors.append(ValidationError(
                    prop, "maxCount",
                    f"Expected at most {max_count} value(s), found {count}",
                    value,
                ))

        # ── Extract scalar for remaining constraints ──
        raw = _extract_raw(value)

        if constraint.get("@required") and raw is None:
            errors.append(ValidationError(prop, "required", f'Property "{prop}" is required'))
            continue

        if raw is None:
            continue

        # Type check
        expected_type = constraint.get("@type")
        if expected_type:
            type_err = _validate_type(raw, expected_type)
            if type_err:
                errors.append(ValidationError(prop, "type", type_err, raw))

        # Numeric (exclude booleans — they are int subclass in Python)
        if "@minimum" in constraint and isinstance(raw, (int, float)) and not isinstance(raw, bool):
            if raw < constraint["@minimum"]:
                errors.append(ValidationError(
                    prop, "minimum",
                    f"Value {raw} below minimum {constraint['@minimum']}", raw,
                ))

        if "@maximum" in constraint and isinstance(raw, (int, float)) and not isinstance(raw, bool):
            if raw > constraint["@maximum"]:
                errors.append(ValidationError(
                    prop, "maximum",
                    f"Value {raw} exceeds maximum {constraint['@maximum']}", raw,
                ))

        # String length
        if "@minLength" in constraint and isinstance(raw, str):
            if len(raw) < constraint["@minLength"]:
                errors.append(ValidationError(
                    prop, "minLength",
                    f"Length {len(raw)} below minimum {constraint['@minLength']}", raw,
                ))

        if "@maxLength" in constraint and isinstance(raw, str):
            if len(raw) > constraint["@maxLength"]:
                errors.append(ValidationError(
                    prop, "maxLength",
                    f"Length {len(raw)} exceeds maximum {constraint['@maxLength']}", raw,
                ))

        # Enumeration
        if "@in" in constraint:
            allowed = constraint["@in"]
            if raw not in allowed:
                errors.append(ValidationError(
                    prop, "in",
                    f"Value {raw!r} not in allowed set {allowed}", raw,
                ))

        # Pattern
        if "@pattern" in constraint and isinstance(raw, str):
            try:
                if not re.search(constraint["@pattern"], raw):
                    errors.append(ValidationError(
                        prop, "pattern",
                        f'"{raw}" does not match pattern "{constraint["@pattern"]}"', raw,
                    ))
            except re.error as exc:
                errors.append(ValidationError(
                    prop, "pattern",
                    f'Invalid regex pattern "{constraint["@pattern"]}": {exc}', raw,
                ))

    return ValidationResult(len(errors) == 0, errors, warnings)


def validate_document(
    doc: dict[str, Any], shapes: Sequence[dict[str, Any]]
) -> ValidationResult:
    """Validate all matching nodes in a document against shapes."""
    all_errors: list[ValidationError] = []
    all_warnings: list[ValidationWarning] = []

    for node in _extract_nodes(doc):
        node_types = _get_types(node)
        for shape in shapes:
            if shape.get("@type") in node_types:
                result = validate_node(node, shape)
                for e in result.errors:
                    e.path = f"{node.get('@id', 'anonymous')}/{e.path}"
                all_errors.extend(result.errors)
                all_warnings.extend(result.warnings)

    return ValidationResult(len(all_errors) == 0, all_errors, all_warnings)


# ── Internal ───────────────────────────────────────────────────────

def _count_values(value: Any) -> int:
    """Count the number of values for cardinality checks.

    - ``None`` (absent property) → 0
    - A list → ``len(list)``
    - Any other single value → 1
    """
    if value is None:
        return 0
    if isinstance(value, list):
        return len(value)
    return 1


def _get_types(node: dict) -> list[str]:
    t = node.get("@type")
    if t is None:
        return []
    return t if isinstance(t, list) else [t]


def _extract_raw(value: Any) -> Any:
    if value is None:
        return None
    if isinstance(value, dict) and "@value" in value:
        return value["@value"]
    if isinstance(value, dict) and not any(k.startswith("@") for k in value):
        return None  # Plain dict without JSON-LD keywords — treat as absent
    if isinstance(value, list) and len(value) > 0:
        return _extract_raw(value[0])
    if isinstance(value, list) and len(value) == 0:
        return None
    return value


def _extract_nodes(doc: Any) -> list[dict]:
    if isinstance(doc, list):
        nodes = []
        for item in doc:
            nodes.extend(_extract_nodes(item))
        return nodes
    if not isinstance(doc, dict):
        return []
    nodes = []
    if "@type" in doc:
        nodes.append(doc)
    if "@graph" in doc:
        nodes.extend(_extract_nodes(doc["@graph"]))
    return nodes


def _validate_type(value: Any, expected: str) -> Optional[str]:
    xsd_type = expected.replace("xsd:", XSD) if expected.startswith("xsd:") else expected
    checks = {
        f"{XSD}string": lambda v: isinstance(v, str),
        f"{XSD}integer": lambda v: isinstance(v, int) and not isinstance(v, bool),
        f"{XSD}double": lambda v: isinstance(v, (int, float)) and not isinstance(v, bool),
        f"{XSD}float": lambda v: isinstance(v, (int, float)) and not isinstance(v, bool),
        f"{XSD}decimal": lambda v: isinstance(v, (int, float)) and not isinstance(v, bool),
        f"{XSD}boolean": lambda v: isinstance(v, bool),
    }
    checker = checks.get(xsd_type)
    if checker and not checker(value):
        short = expected if expected.startswith("xsd:") else xsd_type
        return f"Expected {short}, got {type(value).__name__}: {value}"
    return None

File: packages/js/src/extensions/validation.ts
/**
 * Validation Extensions for JSON-LD (@shape)
 *
 * Provides lightweight, native JSON-LD validation without requiring
 * external SHACL or ShEx tooling. Designed to be simple enough for
 * common use cases while mapping to SHACL for complex scenarios.
 */

import {
  KEYWORD_SHAPE, KEYWORD_REQUIRED, KEYWORD_MINIMUM, KEYWORD_MAXIMUM,
  KEYWORD_MIN_LENGTH, KEYWORD_MAX_LENGTH, KEYWORD_PATTERN,
} from '../keywords';
import {
  ShapeDefinition, PropertyShape, ValidationResult,
  ValidationError, ValidationWarning,
} from '../types';

// ── Shape Validator ───────────────────────────────────────────────

/**
 * Validates a JSON-LD node against a shape definition.
 *
 * @example
 * ```ts
 * const shape = {
 *   "@type": "Person",
 *   "name": { "@required": true, "@type": "xsd:string", "@minLength": 1 },
 *   "email": { "@pattern": "^[^@]+@[^@]+$" },
 *   "age": { "@type": "xsd:integer", "@minimum": 0, "@maximum": 150 },
 * };
 *
 * const result = validateNode(personNode, shape);
 * if (!result.valid) {
 *   console.log(result.errors);
 * }
 * ```
 */
export function validateNode(
  node: any,
  shape: ShapeDefinition
): ValidationResult {
  const errors: ValidationError[] = [];
  const warnings: ValidationWarning[] = [];

  if (node == null || typeof node !== 'object') {
    errors.push({
      path: '.',
      constraint: 'type',
      message: 'Node must be an object',
    });
    return { valid: false, errors, warnings };
  }

  // Type check
  if (shape['@type']) {
    const nodeTypes = getNodeTypes(node);
    if (!nodeTypes.includes(shape['@type'])) {
      errors.push({
        path: '@type',
        constraint: 'type',
        message: `Expected type "${shape['@type']}", found: [${nodeTypes.join(', ')}]`,
        value: nodeTypes,
      });
    }
  }

  // Property constraints
  for (const [prop, constraint] of Object.entries(shape)) {
    if (prop.startsWith('@')) continue; // Skip JSON-LD keywords in shape

    if (typeof constraint !== 'object' || constraint == null) continue;
    const propShape = constraint as PropertyShape;

    const value = node[prop];
    const rawValue = extractRawValue(value);

    // Required check
    if (propShape['@required'] && (rawValue === undefined || rawValue === null)) {
      errors.push({
        path: prop,
        constraint: 'required',
        message: `Property "${prop}" is required`,
      });
      continue; // Skip further checks if missing
    }

    // Skip further validation if value is absent and not required
    if (rawValue === undefined || rawValue === null) continue;

    // Type check
    if (propShape['@type']) {
      const typeError = validateType(rawValue, propShape['@type']);
      if (typeError) {
        errors.push({
          path: prop,
          constraint: 'type',
          message: typeError,
          value: rawValue,
        });
      }
    }

    // Numeric constraints
    if (propShape['@minimum'] !== undefined && typeof rawValue === 'number') {
      if (rawValue < propShape['@minimum']) {
        errors.push({
          path: prop,
          constraint: 'minimum',
          message: `Value ${rawValue} is below minimum ${propShape['@minimum']}`,
          value: rawValue,
        });
      }
    }

    if (propShape['@maximum'] !== undefined && typeof rawValue === 'number') {
      if (rawValue > propShape['@maximum']) {
        errors.push({
          path: prop,
          constraint: 'maximum',
          message: `Value ${rawValue} exceeds maximum ${propShape['@maximum']}`,
          value: rawValue,
        });
      }
    }

    // String length constraints
    if (propShape['@minLength'] !== undefined && typeof rawValue === 'string') {
      if (rawValue.length < propShape['@minLength']) {
        errors.push({
          path: prop,
          constraint: 'minLength',
          message: `String length ${rawValue.length} is below minimum ${propShape['@minLength']}`,
          value: rawValue,
        });
      }
    }

    if (propShape['@maxLength'] !== undefined && typeof rawValue === 'string') {
      if (rawValue.length > propShape['@maxLength']) {
        errors.push({
          path: prop,
          constraint: 'maxLength',
          message: `String length ${rawValue.length} exceeds maximum ${propShape['@maxLength']}`,
          value: rawValue,
        });
      }
    }

    // Pattern constraint
    if (propShape['@pattern'] && typeof rawValue === 'string') {
      const regex = new RegExp(propShape['@pattern']);
      if (!regex.test(rawValue)) {
        errors.push({
          path: prop,
          constraint: 'pattern',
          message: `Value "${rawValue}" does not match pattern "${propShape['@pattern']}"`,
          value: rawValue,
        });
      }
    }
  }

  return { valid: errors.length === 0, errors, warnings };
}

/**
 * Validates an entire JSON-LD document against embedded @shape definitions.
 * Shapes can be defined in the document's @context or alongside nodes.
 */
export function validateDocument(
  doc: any,
  shapes: ShapeDefinition[]
): ValidationResult {
  const allErrors: ValidationError[] = [];
  const allWarnings: ValidationWarning[] = [];

  const nodes = extractNodes(doc);

  for (const node of nodes) {
    const nodeTypes = getNodeTypes(node);

    for (const shape of shapes) {
      if (nodeTypes.includes(shape['@type'])) {
        const result = validateNode(node, shape);
        allErrors.push(...result.errors.map(e => ({
          ...e,
          path: `${node['@id'] ?? 'anonymous'}/${e.path}`,
        })));
        allWarnings.push(...result.warnings);
      }
    }
  }

  return {
    valid: allErrors.length === 0,
    errors: allErrors,
    warnings: allWarnings,
  };
}

// ── Internal Helpers ──────────────────────────────────────────────

function getNodeTypes(node: any): string[] {
  const type = node['@type'];
  if (!type) return [];
  return Array.isArray(type) ? type : [type];
}

function extractRawValue(value: any): any {
  if (value == null) return undefined;
  // JSON-LD expanded value object
  if (typeof value === 'object' && '@value' in value) {
    return value['@value'];
  }
  // Array of values — take first
  if (Array.isArray(value) && value.length > 0) {
    return extractRawValue(value[0]);
  }
  return value;
}

function extractNodes(doc: any): any[] {
  if (Array.isArray(doc)) {
    return doc.flatMap(extractNodes);
  }
  if (doc == null || typeof doc !== 'object') return [];

  const nodes: any[] = [];
  if (doc['@type']) {
    nodes.push(doc);
  }
  if (doc['@graph']) {
    nodes.push(...extractNodes(doc['@graph']));
  }
  return nodes;
}

function validateType(value: any, expectedType: string): string | null {
  const xsdPrefix = 'http://www.w3.org/2001/XMLSchema#';
  const type = expectedType.startsWith('xsd:')
    ? expectedType.replace('xsd:', xsdPrefix)
    : expectedType;

  switch (type) {
    case `${xsdPrefix}string`:
      if (typeof value !== 'string') {
        return `Expected string, got ${typeof value}`;
      }
      break;
    case `${xsdPrefix}integer`:
      if (typeof value !== 'number' || !Number.isInteger(value)) {
        return `Expected integer, got ${typeof value === 'number' ? value : typeof value}`;
      }
      break;
    case `${xsdPrefix}double`:
    case `${xsdPrefix}float`:
    case `${xsdPrefix}decimal`:
      if (typeof value !== 'number') {
        return `Expected number, got ${typeof value}`;
      }
      break;
    case `${xsdPrefix}boolean`:
      if (typeof value !== 'boolean') {
        return `Expected boolean, got ${typeof value}`;
      }
      break;
    case `${xsdPrefix}dateTime`:
      if (typeof value !== 'string' || isNaN(Date.parse(value))) {
        return `Expected ISO 8601 dateTime, got "${value}"`;
      }
      break;
  }
  return null;
}
